subs/pantry/src/Pantry/Types.hs

```diff
@@ -1446,6 +1529,207 @@ rpmToPairs (RawPackageMetadata mname mversion mtree mcabal) = concat
   , maybe [] (\cabal -> ["cabal-file" .= cabal]) mcabal
   ]
 
+parseAndResolvePackageLocation :: Path Abs Dir -> Value -> Parser (IO (NonEmpty PackageLocation))
+parseAndResolvePackageLocation rootDir v = do
+  (Unresolved unresolvedPL) <- parsePackageLocation v
+  pure $ unresolvedPL (Just rootDir)
+
+combineUnresolved :: Unresolved a -> Unresolved b -> Unresolved (a,b)
+combineUnresolved a b = do
+  ua <- a
+  ub <- b
+  pure (ua, ub)
+
+parseSingleObject :: Value -> Parser (Unresolved (PackageLocation, RawPackageLocation))
+parseSingleObject value = withObject "LockFile" (\obj -> do
```

Comment:
why do you add parentheses here and not just go with `$`? Looks a bit unusual (and quite far to the right)